src/common/utils/collection-functions.ts

```diff
@@ -23,6 +24,21 @@ export function getOrInsert<K, V>(map: Map<K, V>, key: K, value: V): V {
   return map.get(key)!;
 }
 
+export async function getOrInsertWithObservable<K>(box: IObservableValue<K | undefined>, getter: () => Promise<K>): Promise<K> {
+  const val = box.get();
+
+  if (val) {
+    return val;
+  }
+
+  const newVal = await getter();
+
+  runInAction(() => box.set(newVal));
```

Comment:
Dangerous. Transaction should be controlled from the useplace.

Reply:
Not when you expect the helper function to be an action